engine/src/main/java/org/terasology/rendering/dag/nodes/CopyImageToHMDNode.java

```diff
@@ -0,0 +1,120 @@
+    private static final ResourceUrn LEFT_EYE_FBO = new ResourceUrn("engine:leftEye");
+    private static final ResourceUrn RIGHT_EYE_FBO = new ResourceUrn("engine:rightEye");
+    // TODO: make these configurable options
+
+    @In
+    private OpenVRProvider vrProvider;
+
+    @In
+    private WorldRenderer worldRenderer;
+
+    @In
+    private Config config;
+
+    @In
+    private ScreenGrabber screenGrabber;
+
+    @In
+    private DisplayResolutionDependentFBOs displayResolutionDependentFBOs;
+
+    private RenderingConfig renderingConfig;
+    private FBO leftEye;
+    private FBO rightEye;
+
+    @Override
+    public void initialise() {
+        renderingConfig = config.getRendering();
+        requiresCondition(() -> (renderingConfig.isVrSupport()
+                && vrProvider != null
+                && vrProvider.isInitialized() ));
+        leftEye = requiresFBO(new FBOConfig(LEFT_EYE_FBO, FULL_SCALE,
+                FBO.Type.DEFAULT).useDepthBuffer(),displayResolutionDependentFBOs);
+        rightEye = requiresFBO(new FBOConfig(RIGHT_EYE_FBO,FULL_SCALE,
+                FBO.Type.DEFAULT).useDepthBuffer(),displayResolutionDependentFBOs);
+        if (this.vrProvider != null) {
+            logger.info("Left eye FBOID:" + Integer.toString(leftEye.fboId));
+            logger.info("Right eye FBOID:" + Integer.toString(leftEye.fboId));
+            logger.info("OpenVR init done.");
+            vrProvider.texType[0].handle = leftEye.colorBufferTextureId;
+            vrProvider.texType[0].eColorSpace = JOpenVRLibrary.EColorSpace.EColorSpace_ColorSpace_Gamma;
+            vrProvider.texType[0].eType = JOpenVRLibrary.EGraphicsAPIConvention.EGraphicsAPIConvention_API_OpenGL;
+            vrProvider.texType[0].write();
+            vrProvider.texType[1].handle = rightEye.colorBufferTextureId;
+            vrProvider.texType[1].eColorSpace = JOpenVRLibrary.EColorSpace.EColorSpace_ColorSpace_Gamma;
+            vrProvider.texType[1].eType = JOpenVRLibrary.EGraphicsAPIConvention.EGraphicsAPIConvention_API_OpenGL;
+            vrProvider.texType[1].write();
+        }
+        addDesiredStateChange(new EnableMaterial("engine:prog.defaultTextured"));
+    }
+
+    @Override
+    public void process() {
+        PerformanceMonitor.startActivity("rendering/copyToVRFrameBuffers");
+        FINAL.bindTexture();
+        if (!renderingConfig.isVrSupport()) {
+            logger.warn("CopyToVRFrameBufferNode processed in non-VR mode.");
+        } else {
+            renderFinalStereoImage(worldRenderer.getCurrentRenderStage());
+        }
+        PerformanceMonitor.endActivity();
+    }
+
+    private void renderFinalStereoImage(RenderingStage renderingStage) {
+        switch (renderingStage) {
+            case LEFT_EYE:
+                vrProvider.updateState();
+                leftEye.bind();
+                glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+                renderFullscreenQuad();
```

Comment:
I'm going to hold off on this for the moment because of the following:

http://stackoverflow.com/questions/3073796/how-to-use-glcopyimage2d

I read this some time back and got the impression that rendering through a passthrough shader to the screen was actually faster than glCopyTexImage2D. I don't know if that's true, but the more concerning part is this:

"Second, as you’ve already observed, glCopyTexImage2D requires its width and height arguments to be power-of-two as well."

So that seems like a non-starter.